### ISSUE TITLE
GCC7 with OpenMP gives incorrect output

### ISSUE THREAD

**Issue by mmcgr** [souffle-lang/souffle#438]
Currently, compiled souffle does not work reliably with gcc7 and openmp. Variables in parallel sections seem to be treated as out of scope. They can be used, but the results are undefined. I'm not sure what's actually happening but if you compile with fsanitize=memory you get out of scope warnings at that location, and gdb doesn't find any symbols. The same code works with earlier gcc and with clang.

In practice, the compiled code may work, or it may give incorrect results.

I'll update this when I have a smaller working sample with the issue.

**Comment by mmcgr** [comment]
Refining this a bit, what's needed is GCC7, OpenMP, and -02/3

**Comment by mmcgr** [comment]
If configure is run with --enable-sanitise-memory then gcc7, clang5 and clang6 all show stack-use-after-scope errors in parallel sections. You can see this with many of the testcases, such as evaluation/access1.
e.g.
```[CC/CXX=suitable compile if needed] ./configure --enable-debug --enable-sanitise-memory```
```SOUFFLE_CATEGORY=Evaluation SOUFFLE_CONFS=-c make check```

Assuming they're not false positives, this may explain incorrect results.

**Comment by mmcgr** [comment]
A small example that gives incorrect output
```.decl A(x0: symbol, x1: symbol)
A("success0", "a").
A("success1", "a").

.decl B(x0: symbol, x1: symbol, x2: symbol)
B("a", "a", "a").

.decl C(x0: symbol)
C(a) :- B(a, _, _).

.decl F(x0: symbol)
F(a) :- C(b), A(a, b).

.output F
```
```souffle -D- -c -pprofile.log test.dl```

With GCC7 and profiling enabled, and no debug mode (to avoid compilation with -O0), F is empty. Otherwise, F has "success0" and "success1". Note that with this example no warnings are given from the address sanitiser.

**Comment by mmcgr** [comment]
Reducing B to 2 attributes gives the expected results.

**Comment by mmcgr** [comment]
More interesting:
```
A("success0", "a").
A("success1", "a").

.decl B(x0: symbol, x1: symbol, x2: symbol)
B("a", "a", "a").

.decl C(x0: symbol)
C(a) :- B(a, _, _).

.output C
```
The correct version will show that C contains 'a'. The incorrect version shows that C contains 'success0'.

**Comment by mmcgr** [comment]
Forcing use of BTree instead of Brie or Auto will also produce the correct output.

**Comment by mmcgr** [comment]
Adding  -fno-tree-pre to the CXXFLAGS in souffle-compile also makes things work.

**Comment by HerbertJordan** [comment]
Hi, thx for the report. This is something I will have a look at asap, it is just quite busy right now.

I hope you can use clang or GCC 6 in the meanwhile.

**Comment by HerbertJordan** [comment]
Hi @mmcgr, I tried to reproduce the issue, but it works for me.

I am compiling the C++ target code manually using
```
g++-7 -std=c++11 -fopenmp target_code.cpp -o target_code_err -I ../include/ -DUSE_LIBZ  -DUSE_SQLITE -lpthread -lsqlite3 -lz -lncurses -O3 -fsanitize=address,leak
```
Can you try and test whether this way it works for you? If not, can you upload the target code generated by your souffle installation?

**Comment by mmcgr** [comment]
It still doesn't work for me, so here's some target code:
[souffle_gcc7.zip](https://github.com/souffle-lang/souffle/files/1206409/souffle_gcc7.zip)

Disabling optimisations for SparseArray::tryUpdateRootInfo got it working for me.
```bool tryUpdateRootInfo(const RootInfoSnapshot& info) __attribute__((optimize("-O0")))```

I tried compiling with 
g++ (GCC) 7.1.1 20170727
and
g++-7 (Ubuntu 7-20170407-0ubuntu2) 7.0.1 20170407 (experimental) [trunk revision 246759]

**Comment by HerbertJordan** [comment]
thx for the code example. I could reproduce it now as well.

An initial investigation revealed that the same issue also occurs without OpenMP, just processing it sequential. 

I tracked down the issue, which is caused by GCC7 being very capable and aggressive in optimizing the code. I am not 100% sure yet how to effectively fix the problem without harming performance.

**Comment by HerbertJordan** [comment]
It seems like GCC7 is producing out-of-thin-air values in connection with the fetch_or call for atomics and relaxed memory order. Technically, this has only been allowed until C++14, but changing the standard flag to this value does not fix it either. This might be a GCC bug or my interpretation might be off ... either way, we need a fix.

@mmcgr can you try replacing the set method in [line 1451 of the Trie.h file](https://github.com/souffle-lang/souffle/blob/master/src/Trie.h#L1451-L1455) to the following code and see whether this fixes the issues you encountered with all your examples? If so, I will clean it up and add it as a fix to the repo.

```Cpp
    bool set(index_type i, op_context& ctxt) {
        atomic_value_t& val = store.getAtomic(i >> LEAF_INDEX_WIDTH, ctxt);
        value_t bit = (1ull << (i & LEAF_INDEX_MASK));

	while(true) {
		auto order = std::memory_order::memory_order_relaxed;

		// load current value
		value_t old = val.load(order);

		// if bit is already set => we are done
		if (old & bit) return false;

		// set the bit, if failed, repeat
		if (!val.compare_exchange_strong(old,old | bit,order,order)) continue;

		// it worked, new bit added
		return true;
	}

    }
```

**Comment by mmcgr** [comment]
All examples I've come up with work fine with the new set method in place. GCC8 had the same problem, and the fix also works with that.
